### ISSUE TITLE
Cloudstack DataSource can't update root password after PR 4997

### ISSUE THREAD

**Issue by benj-n** [canonical/cloud-init#5486]
# Bug report
The PR https://github.com/canonical/cloud-init/pull/4997 introduced a change in the default cache behaviour:
util.del_file() is now called after sources.find_source() in _get_data_source in stages.py and may actually not be called at all.

Considering the following flow:
A single DataSource with network dependency ; after the initial run (which populates obj.pkl), the subsequent runs will:
1/ Not find a DS during "init --local" (because of the dependency)
2/ Not delete obj.pkl (because of PR 4997)
3/ Restore obj.pkl during "init" (with existing = "trust")
4/ Not fetch any "fresh" data from the metadata service

It looks like to me that obj.pkl is not a cache anymore, it's now a persistant representation of an instance configuration that will never ever change.

The issue here is that the Cloudstack DS implements a mechanism to update the root password as described here: https://docs.cloudstack.apache.org/en/latest/adminguide/templates/_cloud_init.html.
This mechanism relies on being able to invalidate the cache and fetch the metadata (new password) on every boot which is not possible anymore.

## Steps to reproduce the problem
As RedHat backported the aforementioned commit in their current cloud-init packages (cloud-init-23.4-7.el9_4.3.0.1.noarch): 
* On a recent cloudstack setup (4.18 and above),
* boot a VM with any EL8 or EL9 flavor (AlmaLinux, Rocky, RHEL)
* Validate root password is the one that has been generated by Cloudstack
* Stop the VM, reset the password via Cloudstack GUI or API, Start the VM
* Validate the root password has not changed after cloud-init has run

## Environment details
- Cloud-init version: all after PR 4997 was merged
- Cloud provider, platform or installer type: Cloudstack Datasource

**Comment by loth** [comment]
@blackboxsw this is another breaking change for CloudStack cases that we are affected by as well. I think you were involved in the older PR about password resets if you could take a look if you had time? Thanks!

**Comment by TheRealFalcon** [comment]
> This mechanism relies on being able to invalidate the cache and fetch the metadata (new password) on every boot which is not possible anymore.

The supported way to achieve this is ensuring that your datasource responds appropriately to the [BOOT update event](https://cloudinit.readthedocs.io/en/latest/explanation/events.html). It is [user configurable](https://cloudinit.readthedocs.io/en/latest/explanation/events.html#example) but can also have hardcoded defaults per datasource.

I submitted [a PR to update the Cloudstack datasource](https://github.com/canonical/cloud-init/pull/5491) to fetch metadata every boot. Please let me know if this solves your issue.

**Comment by benj-n** [comment]
I've commented the draft PR.
But I'll still copy the important part of the comment here:

As far as I understand, as per [1], the metadata will be refreshed if and only if cloud-init is configured to manage the network config.
In the case I'm facing now, we rely on other methods to configure the network so we do have the configuration option:
```
network:
  config: disabled
```
The current code creates a dependency between cloud-init managing the network and being able to refresh the metadata at will. It doesn't look like it can be "fixed" in the DataSource and that could be considered a breaking change ...

[1] [stages.py#L1042](https://github.com/canonical/cloud-init/blob/0af459eac9c670f2a7215ecfe96f4a4b6444eaba/cloudinit/stages.py#L1042)

**Comment by TheRealFalcon** [comment]
Attempt 2 : https://github.com/canonical/cloud-init/pull/5499

Again, let me know if this will work for you.

If not, please also provide logs from `cloud-init collect-logs` so I can get a better handle of exactly what is happening.

**Comment by benj-n** [comment]
@TheRealFalcon 
Just finished testing: https://github.com/canonical/cloud-init/pull/5499 works fine for me.

**Comment by hanisirfan** [comment]
Hi guys. I'm new with cloud-init and just following the integration section on CloudStack documentation. Currently running Rocky Linux 9.4 with `cloud-init.noarch 23.4-7.el9_4.5.0.1` and `CloudStack 4.19.1.1`.  Based on the conversation here, how can I get the reset password running? Tbh I don't know how to custom build packages/test commits etc and I do believe this fix won't be available in the near future to Rocky Linux. Thanks for the help :)

**Comment by benj-n** [comment]
@hanisirfan 

RedHat has published an errata [1] mentioning the fix will be present in "cloud-init-23.4-7.el9_4.6".
In the mean time, we used this workaround that doesn't involve building new/custom packages:
We added a systemd dropin (in `/etc/systemd/system/cloud-final.service.d/delete-object-file.conf`) with the content
```
[Service]
ExecStartPost=/bin/sh -c "/bin/rm -f /var/lib/cloud/instance/obj.pkl || true"
```
It forces cloud-init to create a new obj.pkl file (fresh without any cached password) for every boot. 
It is not ideal, but depending on the case, it can be acceptable as a temporary workaround while waiting for the new packages.


[1] https://access.redhat.com/errata/RHBA-2024:6185

**Comment by benj-n** [comment]
FYI. After RedHat a couple of months ago, this bug has now reached the Ubuntu packages.
https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/2080150

**Comment by hanisirfan** [comment]
@benj-n since I mentioned before we're using Rocky Linux 9.4 with `cloud-init.noarch 23.4-7.el9_4.5.0`. This should include the RedHat you stated on last 2 previous comment, right? Btw what does `obj.pkl` do exactly?

**Comment by holmanb** [comment]
> Btw what does `obj.pkl` do exactly?

@hanisirfan Careful with that - it's mucking about in cloud-init's internals to get it to re-run. I really wouldn't recommend doing this, since it may have unintended side-effects and could break at any point.

Since this is fixed in upstream, it would be best to see when you can get the upstream fix. Have you asked your distro provider when they'll ship 24.3?

**Comment by benj-n** [comment]
@hanisirfan

RedHat errata for EL8 is https://access.redhat.com/errata/RHBA-2024:5301.
It is mentioned that fix is present in package: cloud-init-23.4-7.el8_10.7

RedHat errata for EL9 is https://access.redhat.com/errata/RHBA-2024:6185.
It is mentioned that fix is present in package: cloud-init-23.4-7.el9_4.6

As far as I know, those packages have been made available in Rocky linux repositories a few weeks ago for EL8 and last week for EL9.

**Comment by hanisirfan** [comment]
Sorry if I sound a bit needy. But is there an estimate when cloud-init 24.3.1 will be released for Ubuntu 24.04?

Tbh, I don't know what process did a package maintainer and distro maintainer took to release their updates. Is there anything good that I can read on this? Thanks

**Comment by holmanb** [comment]
> Sorry if I sound a bit needy. But is there an estimate when cloud-init 24.3.1 will be released for Ubuntu 24.04?
> 
> Tbh, I don't know what process did a package maintainer and distro maintainer took to release their updates. Is there anything good that I can read on this? Thanks

You can get it today from the `noble-proposed` pocket.

`rmadison cloud-init` shows this:

```
 cloud-init | 24.2-0ubuntu1~20.04.1              | focal-updates   | source, all
 cloud-init | 24.2-0ubuntu1~22.04.1              | jammy-updates   | source, all
 cloud-init | 24.2-0ubuntu1~24.04.2              | noble-updates   | source, all
 cloud-init | 24.3.1-0ubuntu0~20.04.1            | focal-proposed  | source, all
 cloud-init | 24.3.1-0ubuntu0~22.04.1            | jammy-proposed  | source, all
 cloud-init | 24.3.1-0ubuntu0~24.04.1            | noble-proposed  | source, all
```
Once it completes SRU verification, it will be available in `noble-updates`.

See more on testing proposed packages [here](https://docs.cloud-init.io/en/latest/howto/ubuntu_test_prerelease.html#test-pre-release-cloud-init).

You can track the SRU status and report bugs against [the new package here](https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/2079224). @hanisirfan if this package does not resolve the issue for you, please report it too that bug so that we can resolve the issue as soon as we can.
